Make ActiveAdmin Rails 6 compatible

ionut998 opened the pull request:
Opening this PR to get some opinions and ideas about making ActiveAdmin Rails 6 compatible
I used a few unreleased versions of gems that are compatible with Rails 6, but I will change to the official release once that's released.
Most of the tests pass and I am trying now to fix the rest. I tested an existent application against this version of ActiveAdmin and it looks like everything works fine.

Closes #5558.

deivid-rodriguez commented on Gemfile:
```diff
@@ -22,7 +22,9 @@ group :docs do
+# Use the unreleased version of inherited_resources because is compatible with Rails 6
+gem 'inherited_resources', github: 'activeadmin/inherited_resources'
+gem "rails", "6.0.0.beta2"
+gem "activerecord-jdbcsqlite3-adapter", github: 'jruby/activerecord-jdbc-adapter', platform: :jruby
```
I think we can follow the same strategy as in https://github.com/activeadmin/inherited_resources/pull/479. Let's leave the main Gemfile intact for now, and tackle Rails 6.0 on a different gemfile, for example, `gemfiles/rails_60.gemfile{,.lock}`.

ionut998 replied:
I separated the Gemfile for Rails 6 @deivid-rodriguez

yagudaev commented:
@ionut998 great work thanks for the PR it is much needed.

I tried using your branch and hit an issue when creating generators with rails 6 as follow:

```
my-app (scaffold-models) $ rails g model User first_name:string last_name:string email:string 
/Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.1/lib/bootsnap/load_path_cache/core_ext/active_support.rb:76:in `block in load_missing_constant': uninitialized constant ActionController::Responder (NameError)
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.1/lib/bootsnap/load_path_cache/core_ext/active_support.rb:8:in `without_bootsnap_cache'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.1/lib/bootsnap/load_path_cache/core_ext/active_support.rb:76:in `rescue in load_missing_constant'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.1/lib/bootsnap/load_path_cache/core_ext/active_support.rb:58:in `load_missing_constant'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/inherited_resources-1.3.1/lib/inherited_resources/responder.rb:2:in `<module:InheritedResources>'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/inherited_resources-1.3.1/lib/inherited_resources/responder.rb:1:in `<main>'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.1/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:21:in `require'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.1/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:21:in `block in require_with_bootsnap_lfi'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.1/lib/bootsnap/load_path_cache/loaded_features_index.rb:83:in `register'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.1/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:20:in `require_with_bootsnap_lfi'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.1/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:29:in `require'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/inherited_resources-1.3.1/lib/inherited_resources.rb:4:in `<main>'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.1/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:21:in `require'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.1/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:21:in `block in require_with_bootsnap_lfi'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.1/lib/bootsnap/load_path_cache/loaded_features_index.rb:83:in `register'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.1/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:20:in `require_with_bootsnap_lfi'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.1/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:29:in `require'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/bundler/gems/activeadmin-d935f2b13efc/lib/active_admin.rb:9:in `<main>'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.1/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:21:in `require'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.1/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:21:in `block in require_with_bootsnap_lfi'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.1/lib/bootsnap/load_path_cache/loaded_features_index.rb:83:in `register'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.1/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:20:in `require_with_bootsnap_lfi'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.1/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:29:in `require'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/bundler/gems/activeadmin-d935f2b13efc/lib/activeadmin.rb:1:in `<main>'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.1/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:21:in `require'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.1/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:21:in `block in require_with_bootsnap_lfi'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.1/lib/bootsnap/load_path_cache/loaded_features_index.rb:83:in `register'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.1/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:20:in `require_with_bootsnap_lfi'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.1/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:29:in `require'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/bundler-2.0.1/lib/bundler/runtime.rb:81:in `block (2 levels) in require'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/bundler-2.0.1/lib/bundler/runtime.rb:76:in `each'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/bundler-2.0.1/lib/bundler/runtime.rb:76:in `block in require'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/bundler-2.0.1/lib/bundler/runtime.rb:65:in `each'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/bundler-2.0.1/lib/bundler/runtime.rb:65:in `require'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/bundler-2.0.1/lib/bundler.rb:114:in `require'
        from /Users/mike/projects/my/my-app/config/application.rb:20:in `<top (required)>'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/spring-2.0.2/lib/spring/application.rb:92:in `require'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/spring-2.0.2/lib/spring/application.rb:92:in `preload'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/spring-2.0.2/lib/spring/application.rb:153:in `serve'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/spring-2.0.2/lib/spring/application.rb:141:in `block in run'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/spring-2.0.2/lib/spring/application.rb:135:in `loop'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/spring-2.0.2/lib/spring/application.rb:135:in `run'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/spring-2.0.2/lib/spring/application/boot.rb:19:in `<top (required)>'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/2.6.0/rubygems/core_ext/kernel_require.rb:54:in `require'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/2.6.0/rubygems/core_ext/kernel_require.rb:54:in `require'
        from -e:1:in `<main>'
```

deivid-rodriguez commented on activeadmin.gemspec:
```diff
@@ -22,10 +22,10 @@ Gem::Specification.new do |s|
   s.add_dependency 'formtastic', '~> 3.1'
   s.add_dependency 'formtastic_i18n', '~> 0.4'
-  s.add_dependency 'inherited_resources', '~> 1.7'
+  s.add_dependency 'inherited_resources'
```
Is this change needed? Master branch of inherited resources is still compatible with this requirement I think.

ionut998 replied:
yes, this is actually not needed
I added it back

deivid-rodriguez commented on activeadmin.gemspec:
```diff
@@ -22,10 +22,10 @@ Gem::Specification.new do |s|
   s.add_dependency 'jquery-rails', '~> 4.2'
   s.add_dependency 'kaminari', '~> 1.0', '>= 1.0.1'
-  s.add_dependency 'railties', '>= 5.0', '< 5.3'
+  s.add_dependency 'railties', '>= 5.0'
```
Let's keep an upper limit here. I think `< 6.0` to allow 6.0 prereleases works for now.

ionut998 replied:
I changed to keep the upper limit as `< 6.0`

deivid-rodriguez commented:
@yagudaev You need to manually add `gem "inherited_resources", git: "https://github.com/activeadmin/inherited_resources"` to your `Gemfile` for now. Otherwise `bundler` will resolve `inherited_resources` to a very old version (which had no requirement on the rails version) and you'll get that weird error.

deivid-rodriguez commented on gemfiles/rails_60.gemfile:
```diff
@@ -0,0 +1,12 @@
+gem "rails", "6.0.0.beta2"
+
+# Use the unreleased version of gems because they are compatible with Rails 6
+gem 'inherited_resources', github: 'activeadmin/inherited_resources'
```
```suggestion
gem 'inherited_resources', git: 'https://github.com/activeadmin/inherited_resources'
```

To skip bundler warnings.

deivid-rodriguez commented on gemfiles/rails_60.gemfile:
```diff
@@ -0,0 +1,12 @@
+
+# Use the unreleased version of gems because they are compatible with Rails 6
+gem 'inherited_resources', github: 'activeadmin/inherited_resources'
+gem "activerecord-jdbcsqlite3-adapter", github: 'jruby/activerecord-jdbc-adapter', platform: :jruby
```
```suggestion
gem "activerecord-jdbcsqlite3-adapter", git: 'https://github.com/jruby/activerecord-jdbc-adapter', platform: :jruby
```

To skip more bundler warnings.

deivid-rodriguez commented on gemfiles/rails_60.gemfile:
```diff
@@ -0,0 +1,12 @@
+# Use the unreleased version of gems because they are compatible with Rails 6
+gem 'inherited_resources', github: 'activeadmin/inherited_resources'
+gem "activerecord-jdbcsqlite3-adapter", github: 'jruby/activerecord-jdbc-adapter', platform: :jruby
+gem 'sqlite3', '~> 1.3.6', platform: :mri, group: :test
```
Remove this line since you have already locked the dependency in `Gemfile.common`.

deivid-rodriguez commented:
By the way, I don't know why CircleCI logs are "swallowing" cucumber errors. Something to do with the `parallel_tests` setup. If you run the failing features locally you'll see the actual errors.

ionut998 commented:
That last test seem like a tricky one

The actual error is `Invalid single-table inheritance type: Publisher is not a subclass of User (ActiveRecord::SubclassNotFound)` from [this test](https://github.com/activeadmin/activeadmin/blob/master/features/comments/commenting.feature#L134)

Interestingly if I only run this test then it passes. Something is being set by the [test before](https://github.com/activeadmin/activeadmin/blob/master/features/comments/commenting.feature#L118) that make it fail. If I delete the `Commenting on a STI superclass` test then the `Commenting on a STI subclass` test passes.

If I follow the trace and look into Rails it leads me to [this comparison](https://github.com/rails/rails/blob/6a1a1e66ea7a917942bd8369fa8dbfedce391dab/activerecord/lib/active_record/inheritance.rb#L242) that returns false when it shouldn't 
These are the values when this comparison happens:
```
self = User(id: integer, type: string...
subclass = Publisher(id: integer, type: string...
descendants=[User::VIP(id: integer, type: string...), Publisher(id: integer, type: string...)]
```
and strangely when I do `descendants.include?(subclass)` I get false

Any ideas @deivid-rodriguez

deivid-rodriguez commented:
@ionut998 Seems like a code (re)loading issue... :thinking:. Rails 6 will introduce [a new code loader](https://github.com/fxn/zeitwerk) (which looks awesome, by the way). Maybe you can try setting `config.autoloader = :classic` in the test application. If it passes, we know it's related to the new loader.

ionut998 commented:
Thanks @deivid-rodriguez 
Setting the autoloader to `classic` fixes the issue
I followed the documentation from [here](https://weblog.rubyonrails.org/2019/2/22/zeitwerk-integration-in-rails-6-beta-2/#stis) and added the extra initializer to preload the leafs of the STI but I still get that error.

deivid-rodriguez commented:
To get the `upload_coverage` job green, you need to adjust the total never of jobs here (sum one to the 19 per job being added in this PR):

https://github.com/activeadmin/activeadmin/blob/17988a9ecd85b144529501dbf2a0c0e9ac09c53b/.circleci/config.yml#L65

deivid-rodriguez commented on .circleci/config.yml:
```diff
@@ -208,6 +208,18 @@ jobs:
+
+    steps:
+      *test_app_steps
+
```
Double blank line :smiley:

ionut998 replied:
I have removed the extra line

deivid-rodriguez commented on spec/support/rails_template.rb:
```diff
@@ -114,6 +114,8 @@ class Tagging < ActiveRecord::Base
 
 RUBY
 
+inject_into_file 'config/environments/test.rb', "\n  config.autoloader = :classic\n", after: 'Rails.application.configure do' if Rails::VERSION::MAJOR >= 6
```
Can we add a TODO note about fixing this in the future, maybe with a reference to the STI cucumber feature that fails with `zeitwerk`?

ionut998 replied:
I added the TODO note

deivid-rodriguez commented:
@ionut998 Thanks for doing this! If you can update to the recently released 6.0.0.beta3 since we are at it, other than that it looks great to me!

ionut998 commented:
Changed to Rails 6 beta3 @deivid-rodriguez

javierjulio commented on .circleci/config.yml:
```diff
@@ -62,7 +62,7 @@ version: 2.1
     name: Upload coverage results to Code Climate
     command: |
-      coverage/test-reporter sum-coverage coverage/codeclimate.*.json --parts 19 --output coverage/codeclimate.total.json
+      coverage/test-reporter sum-coverage coverage/codeclimate.*.json --parts 22 --output coverage/codeclimate.total.json
```
@ionut998 would you mind adding a comment above this? @deivid-rodriguez I saw your comment about this change and didn't remember what it was, thanks. Is the parts number the total count of builds/jobs plus 1? Is the plus 1 the lint job? Maybe a note like:

```
# The parts number is the total jobs count plus 1 (for the lint job)
coverage/test-reporter sum-coverage coverage/codeclimate.*.json --parts 22 --output coverage/codeclimate.total.json
```

deivid-rodriguez replied:
Yeah, that's it. It's the number of jobs that generate coverage reports.

ionut998 replied:
Added the comment

ionut998 commented:
Happy to help @javierjulio

deivid-rodriguez commented:
@ionut998 Sorry for the many requests. I noticed that with the upgrade to beta3, also `zeitwerk` got upgraded to `1.3.3`. Who knows whether this bug fix release could fix our problems with the new loader... Mind giving it a try?

ionut998 commented:
No problem @deivid-rodriguez 
I actually tried before pushing the upgrade. I removed the `autoloader = :classic` from the rails6_beta3 test application and added the custom initializer for preloading the STI leafs but it still errored in the same way as before.

javierjulio commented:
@ionut998 thanks again for doing this! We all really appreciate it. ❤️

deivid-rodriguez commented:
Funny that the commit number 6000 in the repo has added support for Rails 6.0 :smiley:.

javierjulio commented:
Wow, you're right! So cool. Let the record show!

<img width="459" alt="activeadmin-6000-commits-for-rails6-support" src="https://user-images.githubusercontent.com/198547/54390503-03c8f700-4679-11e9-98f8-f45ab89671d2.png">

ionut998 commented:
Perfect timing for Rails 6 support 😃
